File: binary_search/744_find_smallest_letter_greater_than_target.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        left, right = 0, len(letters) - 1

        while left <= right:
            mid = (left + right) // 2

            if letters[mid] > target:
                right = mid - 1
            elif letters[mid] <= target:
                left = mid + 1

        return letters[0] if left == len(letters) else letters[left]
        

File: binary_search/744_find_smallest_letter_greater_than_target.java
class Solution {
    public char nextGreatestLetter(char[] letters, char target) {
        int left = 0, right = letters.length - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;

            if (letters[mid] > target) {
                right = mid;
            } else if (letters[mid] <= target) {
                left = mid + 1;
            }
        }
        if (letters[left] > target) {
            return letters[left];
        }
        return letters[0];
    }
}